test/vm_tests/transformer/declaration_parsing_test.dart

```diff
@@ -222,14 +285,28 @@ main() {
           expect(declarations.abstractState, hasLength(2));
 
           expect(declarations.abstractState[0].node.name.name, 'AbstractFooState1');
+          expect(declarations.abstractState[0].companionNode, isNull);
           expect(declarations.abstractState[1].node.name.name, 'AbstractFooState2');
+          expect(declarations.abstractState[1].companionNode, isNull);
           expect(declarations.abstractState[0].meta, new isInstanceOf<annotations.AbstractState>());
           expect(declarations.abstractState[1].meta, new isInstanceOf<annotations.AbstractState>());
 
           expectEmptyDeclarations(abstractState: false);
           expect(declarations.declaresComponent, isFalse);
         });
 
+        test('abstract state class with builder-compatible dual-class setup', () {
+          setUpAndParse('''
+            @AbstractState() class _\$AbstractFooState {}
+            class AbstractFooState {}
+          ''');
+
+          expect(declarations.abstractState, hasLength(1));
+          expect(declarations.abstractState[0].node?.name?.name, '_\$AbstractFooState');
+          expect(declarations.abstractState[0].companionNode?.name?.name, 'AbstractFooState');
+          expect(declarations.abstractState[0].meta, new isInstanceOf<annotations.AbstractState>());
+        });
+
```

Comment:
Props and State mixins don't have companion nodes, we're not renaming them to `_$`-prefixed versions until after we drop Dart 1 support because we can't just extend them.